### PR TITLE
Agnhost: make it possible to pass the addresses to listen on for udp

### PR DESCRIPTION
**Pull request by fedepaol**
**What type of PR is this?**

/kind failing-test

**What this PR does / why we need it**:

The current udp implementation listens on any for tcp, udp and tcp. There are some cases where it makes sense to listen on specific addresses (especially udp, see https://github.com/kubernetes/kubernetes/issues/95565).
This is because UDP is connectionless, and in order to conntrack to work, the application must ensure that the src of the reply is the same as the dest of the request. The easiest way to do that is to bind explicitly on an ip.
Here we pass an optional parameter that contains a comma separated list of addresses. 

The parameter works only for UDP, since it's the only meaningful case that needs to be addressed, and extending the other protocols would have made the code unnecessary complex for no reason.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/images/agnhost/netexec/netexec.go [hunk 1]
```diff
@@ -123,6 +127,7 @@ func init() {
 	CmdNetexec.Flags().IntVar(&udpPort, "udp-port", 8081, "UDP Listen Port")
 	CmdNetexec.Flags().IntVar(&sctpPort, "sctp-port", -1, "SCTP Listen Port")
 	CmdNetexec.Flags().StringVar(&httpOverride, "http-override", "", "Override the HTTP handler to always respond as if it were a GET with this path & params")
+	CmdNetexec.Flags().StringVar(&bindToIPs, "udp-bind-on", "", "A comma separated list of ip addresses to bind the udp server listeners to")
```
Comment (aojea) [inline test/images/agnhost/netexec/netexec.go:130]: naming is hard `udp-listen-addresses`, not attached to it, just want to hear other thoughts

File: test/images/agnhost/netexec/netexec.go [hunk 2]
```diff
@@ -192,7 +199,6 @@ func addRoutes(mux *http.ServeMux, exitCh chan shutdownRequest) {
 }
 
 func startServer(server *http.Server, exitCh chan shutdownRequest, fn func() error) {
-	log.Printf("Started HTTP server on port %d", httpPort)
```
Comment (aojea) [inline test/images/agnhost/netexec/netexec.go:195]: any special reason to remove it?
actually, I think we should add logging too to the udp servers when starting

Comment (aojea) [reply test/images/agnhost/netexec/netexec.go:195]: duh, it does have logging, forget it, just enable this one if there is no special reason to remove it

Comment (fedepaol) [reply test/images/agnhost/netexec/netexec.go:195]: My bad, must have been for when I tried to extend the multi-address to other protocols.
Bringing it back.

File: test/images/agnhost/netexec/netexec.go (hunk 1 above)
Comment (fedepaol) [reply test/images/agnhost/netexec/netexec.go:130]: I don't have strong opinions, I found it quite hard to find the proper name :-)

File: test/images/agnhost/netexec/netexec.go [hunk 3]
```diff
@@ -37,19 +37,23 @@ import (
 	"github.com/spf13/cobra"
 
 	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/gengo/examples/set-gen/sets"
```
Comment (aojea) [inline test/images/agnhost/netexec/netexec.go:40]:
```suggestion
	"k8s.io/apimachinery/pkg/util/sets"
```

File: test/images/agnhost/netexec/netexec.go [hunk 4]
```diff
@@ -642,3 +649,21 @@ func assertNoError(err error, detail string) {
...
+	split := strings.Split(addresses, ",")
+	for _, address := range split {
+		netAddr := net.ParseIP(address)
+		if netAddr == nil {
+			log.Printf("Invalid address %s", address)
+		}
+		res = append(res, address)
+	}
+	set := sets.NewString(res...)
+	return set.List()
```
Comment (aojea) [inline test/images/agnhost/netexec/netexec.go:668]: set := sets.NewString(res...)
	split := strings.Split(addresses, ",")
	for _, address := range split {
		netAddr := net.ParseIP(address)
		if netAddr == nil {
			log.Printf("Invalid address %s", address)
                        # or we fail here to avoid failing later
                        # or we swallow it with a continue
		}
		set.Insert(address)
	}
	return set.List()

File: test/images/agnhost/netexec/netexec.go (hunk 3 above)
Comment (fedepaol) [reply test/images/agnhost/netexec/netexec.go:40]: I should never trust vscode when I choose a package

Comment (fedepaol) [reply test/images/agnhost/netexec/netexec.go:40]: Done

File: test/images/agnhost/netexec/netexec.go (hunk 4 above)
Comment (fedepaol) [reply test/images/agnhost/netexec/netexec.go]: Done, went with the swallowing

**Comment by aojea** [comment]
/ok-to-test
you´ll need to bump the agnhost version, we should coordinate somehow to not collide with https://github.com/kubernetes/kubernetes/pull/98639

**Comment by aojea** [comment]
/assign @MrHohn @uablrek 
PTAL the context is here https://github.com/kubernetes/kubernetes/issues/95565, if you don´t bind the udp servers it takes a different ip address to answer, so the NAT is never reversed

**Comment by fedepaol** [comment]
I see this failing because I added the `sets` import.
Does it make sense to add it to the `BUILD` file or should I go with go maps?

**Comment by aojea** [comment]
`hack/update-bazel.sh` will do the magic

File: test/images/agnhost/netexec/netexec.go [hunk 5]
```diff
@@ -642,3 +649,22 @@ func assertNoError(err error, detail string) {
...
+	for _, address := range split {
+		netAddr := net.ParseIP(address)
+		if netAddr == nil {
+			log.Printf("Invalid address %s", address)
+			continue
+		}
+		res = append(res, address)
+	}
+	set := sets.NewString(res...)
+	return set.List()
```
Comment (aojea) [inline test/images/agnhost/netexec/netexec.go:669]: what happens if all addresses are invalid?
I have second thoughts about swallowing instead of failing, it seems failing is safer and we avoid false positives, i.e. someone puts a bad IP, it listen in all interfaces, the test passes but we are not testing the IP that we want

Comment (fedepaol) [reply test/images/agnhost/netexec/netexec.go:669]: If all the addresses are invalid, we are gonna return an empty slice and we won't be listening on udp.
The test would fail, but I do agree, better be explicit. It would also make our life easier when triaging the failure.

**Comment by uablrek** [comment]
> PTAL the context is here #95565, if you don´t bind the udp servers it takes a different ip address to answer, so the NAT is never reversed

Sorry for the late reply. @aojea I am a bit unsure what I should do but I can confirm that UDP to a service with endpoints in hostNetwork works if the server binds to the podIPs only.

But I think this "the podIPs only" is important since those will become the endpoint addresses. From v1.20 it is possible to have multiple **node IPs** even in private clouds and those will become **podIPs** for a pod in hostNetwork for dual stack;
```
$ kubectl get node vm-002 -o json | jq .status.addresses
[
  {
    "address": "192.168.1.2",
    "type": "InternalIP"
  },
  {
    "address": "1000::1:c0a8:102",
    "type": "InternalIP"
  },
  {
    "address": "vm-002",
    "type": "Hostname"
  }
]
$ kubectl get pod mserver-hostnet-scb9n -o json | jq .status.podIPs
[
  {
    "ip": "192.168.1.2"
  },
  {
    "ip": "1000::1:c0a8:102"
  }
]
```
@fedepaol good thinking to allow multiple addresses to dual-stack can be supported.

**Comment by aojea** [comment]
@uablrek , I just wanted your opinion, you always have good insights in this complex networking scenarios. Thanks

@fedepaol I think that is missing the release-node field and bumping the agnhost version in
build/dependencies.yaml
test/images/agnhost/VERSION
test/images/agnhost/agnhost.go

**Comment by fedepaol** [comment]
> @uablrek , I just wanted your opinion, you always have good insights in this complex networking scenarios. Thanks
> 
> @fedepaol I think that is missing the release-node field and bumping the agnhost version in
> build/dependencies.yaml
> test/images/agnhost/VERSION
> test/images/agnhost/agnhost.go

Yes, was not sure we wanted to do this contextually to this pr

**Comment by fedepaol** [comment]
@aojea done

**Comment by aojea** [comment]
> @aojea done

have you tested it locally? otherwise I will try to do it later, better to be sure before merging, change these images is costly

**Comment by fedepaol** [comment]
> > @aojea done
> 
> have you tested it locally? otherwise I will try to do it later, better to be sure before merging, change these images is costly

Yes, I understand that. I used it to test https://github.com/kubernetes/kubernetes/pull/98884 locally, hope I did not miss anything

File: test/images/agnhost/netexec/netexec.go [hunk 6]
```diff
@@ -174,7 +186,6 @@ func main(cmd *cobra.Command, args []string) {
 		startServer(server, exitCh, server.ListenAndServe)
 	}
 }
-
```
Comment (claudiubelu) [inline test/images/agnhost/netexec/netexec.go:177]: unrelated change.

**Review by claudiubelu — COMMENTED** [review]
/lgtm

File: test/images/agnhost/netexec/netexec.go (hunk 6 above)
Comment (fedepaol) [reply test/images/agnhost/netexec/netexec.go:177]: Fixed

File: test/images/agnhost/netexec/netexec.go [hunk 7]
```diff
@@ -162,7 +167,14 @@ func main(cmd *cobra.Command, args []string) {
...
+	udpBindTo, err := parseAddresses(udpListenAddresses)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for _, address := range udpBindTo {
+		go startUDPServer(address, udpPort)
+	}
 	if sctpPort != -1 {
 		go startSCTPServer(sctpPort)
```
Comment (aojea) [inline test/images/agnhost/netexec/netexec.go:179]: I think that SCTP is stateless and will have the same issue, no?

at least here is failing because only the pod from the other node replies, but not in the same node
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-dualstack-azure-e2e-master-iptables/1362333391973781504

File: build/dependencies.yaml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 dependencies:
   # agnhost: bump this one first
   - name: "agnhost"
-    version: "2.27"
+    version: "2.28"
```
Comment (aojea) [inline build/dependencies.yaml:4]: something is wrong, seems current published agnhost image is 2.28

 gcr.io/k8s-staging-e2e-test-images/agnhost@sha256:4f3eec1e602e1e0b0e5858c8f3d399d0ef536799ec1f7666e025f211001cb706

Comment (aojea) [reply build/dependencies.yaml:4]: cc @claudiubelu 
https://github.com/kubernetes/kubernetes/blob/033d950152dc83fcd9afb546f8c112feb9e2c6f7/build/dependencies.yaml#L4-L13

**Comment by aojea** [comment]
/lgtm
I need an approval so we can fix the latest reds in sig-network testgrid
/cc @thockin @aramase 
https://testgrid.k8s.io/sig-network-dualstack-azure-e2e

**Comment by BenTheElder** [comment]
/approve
see the kubernetes dev notice, you'll need to rebase / hack/update-bazel.sh in the rebase 😅

**Comment by fedepaol** [comment]
I rebased. Not sure what is the issue with `pull-kubernetes-dependencies` / `pull-kubernetes-verify`.

I tried to run locally the failing commands (`hack/verify-vendor.sh` on one side, `/hack/update-internal-modules.sh` on the other, but nothing changed on my repo and they did not seem to fail.

**Comment by aojea** [comment]
> I rebased. Not sure what is the issue with `pull-kubernetes-dependencies` / `pull-kubernetes-verify`.
> 
> I tried to run locally the failing commands (`hack/verify-vendor.sh` on one side, `/hack/update-internal-modules.sh` on the other, but nothing changed on my repo and they did not seem to fail.

there is other pr trying to fix it

**Comment by fedepaol** [comment]
Fixed the dependencies errors

**Comment by aojea** [comment]
> @fedepaol: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-kubernetes-unit	[f6451c9](https://github.com/kubernetes/kubernetes/commit/f6451c93b132236161a17eabd1584cf490848d6a)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98883/pull-kubernetes-unit/1367785835595829248)	`/test pull-kubernetes-unit`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=98883). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Afedepaol). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

heh, this is crazy, this PR hit all the roadblacks, seems an unit test is crazy
> TestFirewallObject

**Comment by aojea** [comment]
/hold
let me hold this until the revert fixing the units failures merges

**Comment by BenTheElder** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/post-kubernetes-push-e2e-agnhost-test-images/1368101986561429504

Postsubmit failed. I'm about to log off for now but probably worth looking into at some point

**Comment by aojea** [comment]
> https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/post-kubernetes-push-e2e-agnhost-test-images/1368101986561429504
> 
> Postsubmit failed. I'm about to log off for now but probably worth looking into at some point

I re triggered

**Comment by aojea** [comment]
```
make[1]: Leaving directory '/workspace/test/images/agnhost'
/workspace/_tmp/test-images-build.BEFmoj /workspace/test/images
Registering qemu-*-static binaries in the kernel
./image-util.sh: line 168: sudo: unbound variable
make: *** [Makefile:43: all-build-and-push] Error 1
ERROR
ERROR: build step 0 "gcr.io/k8s-testimages/gcb-docker-gcloud:v20201130-750d12f" failed: step exited with non-zero status: 2
ERROR: (gcloud.builds.submit) build ee9ce321-9f2c-45db-9196-8e4a60276719 completed with status "FAILURE"
```

**Comment by aojea** [comment]
https://github.com/kubernetes/kubernetes/pull/99889
